File: src/main/java/net/usbwire/usbplus/commands/CompassCommand.java
package net.usbwire.usbplus.commands;

import net.usbwire.usbplus.USBPlus;
import net.usbwire.usbplus.config.Config;
import net.usbwire.usbplus.features.Compass;
import net.usbwire.usbplus.util.chat.Coordinates;
import net.usbwire.usbplus.util.Util;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.incendo.cloud.annotations.Command;
import org.incendo.cloud.annotations.processing.CommandContainer;

@CommandContainer
public final class CompassCommand {
	private static final Pattern coordinatesRegex =
			Pattern.compile("\\(?(-?\\d{1,4}),? (-?\\d{1,4}),? (-?\\d{1,4})\\)?");

	public CompassCommand() {
		// super("compass");
	}

	@Command("compass")
	public void handle() {
		Compass.createCompass();
	}

	@Command("compass help")
	public void handleHelp() {
		Util.chat("Usage for /compass:\n"
				+ "/compass - Generate a JM/Xaero waypoint for the position your compass is currently pointing at\n"
				+ "/compass pos <x> <y> <z> - Generate a JM/Xaero waypoint for these coordinates\n"
				+ "/compass wiki (x, y, z) - Generate a JM/Xaero waypoint for wiki-formatted coordinates");
	}

	@Command("compass pos <x> <y> <z>")
	public void handlePos(int x, int y, int z) {
		Coordinates.Coordinate coordinates = new Coordinates.Coordinate(x, y, z);
		Compass.createCompass(coordinates);
	}

	@Command("compass wiki <input>")
	public void handleWiki(String input) {
		Matcher matcher = coordinatesRegex.matcher(input);
		if (!matcher.matches()) {
			Util.chat("Invalid String! Should be formatted like 'x, y, z' or '(x, y, z)'");
			return;
		}
		matcher.find();
		int x = Integer.parseInt(matcher.group(1));
		int y = Integer.parseInt(matcher.group(2));
		int z = Integer.parseInt(matcher.group(3));
		Coordinates.Coordinate coordinates = new Coordinates.Coordinate(x, y, z);
		Compass.createCompass(coordinates);
	}
}

File: src/main/kotlin/net/usbwire/usbplus/commands/CompassCommand.kt
package net.usbwire.usbplus.commands

import gg.essential.api.commands.*
import gg.essential.api.utils.GuiUtil
import net.usbwire.usbplus.USBPlus
import net.usbwire.usbplus.config.Config
import net.usbwire.usbplus.features.Compass
import net.usbwire.usbplus.util.chat.Coordinates
import net.usbwire.usbplus.util.Util

object CompassCommand : Command("compass") {
	@DefaultHandler
	fun handle () {
		Compass.createCompass()
	}

	@SubCommand("help")
	fun handleHelp () {
		Util.chat("""Usage for /compass:
		|/compass - Generate a JM/Xaero waypoint for the position your compass is currently pointing at
		|/compass pos <x> <y> <z> - Generate a JM/Xaero waypoint for these coordinates
		|/compass wiki (x, y, z) - Generate a JM/Xaero waypoint for wiki-formatted coordinates
		""".trimMargin("|"))
	}

	@SubCommand("pos")
	fun handlePos (x: Int, y: Int, z: Int) {
		val coordinates = Coordinates.Coordinates(x, y, z)
		Compass.createCompass(coordinates)
	}

	val coordinatesRegex = """\(?(-?[\d]{1,4}),? (-?[\d]{1,4}),? (-?[\d]{1,4})\)?""".toRegex()
	@SubCommand("wiki")
	fun handleWiki (@Greedy input: String) {
		if (!coordinatesRegex.matches(input)) {
			Util.chat("Invalid String! Should be formatted like 'x, y, z' or '(x, y, z)'")
			return
		}
		val result = coordinatesRegex.find(input)
		val (x, y, z) = result!!.destructured

		val coordinates = Coordinates.Coordinates(x.toInt(), y.toInt(), z.toInt())
		Compass.createCompass(coordinates)
	}
}
